src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable.js

```diff
@@ -65,31 +53,45 @@ export class VisualizeEmbeddable extends Embeddable  {
         this.uiState.set(key, customization[key]);
       });
       this.customization = customization;
-      isDirty = true;
       this.uiState.on('change', this._uiStateChangeHandler);
     }
+  }
 
+  /**
+   * Retrieve the panel title for this panel from the container state.
+   * This will either return the overwritten panel title or the visualization title.
+   */
+  getPanelTitle(containerState) {
     let derivedPanelTitle = '';
     if (!containerState.hidePanelTitles) {
       derivedPanelTitle = containerState.customTitle !== undefined ?
         containerState.customTitle :
         this.savedVisualization.title;
     }
+    return derivedPanelTitle;
+  }
+
+  onContainerStateChanged(containerState) {
+    this.transferCustomizationsToUiState(containerState);
+
+    const updatedParams = {};
+
+    // Check if timerange has changed
```

Comment:
Basically, what we're doing here is:

- manual diff
- deriving properties based on that diff

If we're doing this sort of thing often, might be worth looking into generalizing this somehow.